### PR TITLE
mon/MonClient: ENXIO when sending command to down mon

### PR DESCRIPTION
**Pull request by liewegas**
Fixes: https://tracker.ceph.com/issues/40792

### CODE REVIEW

**Comment by tchaikov** [comment]
jenins, retest this please

File: src/mon/MonClient.cc [hunk 1]
```diff
@@ -1154,9 +1167,12 @@ void MonClient::start_mon_command(const std::vector<string>& cmd,
                                   ceph::buffer::list *outbl, string *outs,
                                   Context *onfinish)
 {
+  ldout(cct,10) << __func__ << " cmd=" << cmd << dendl;
   std::lock_guard l(monc_lock);
   if (!initialized || stopping) {
-    onfinish->complete(-ECANCELED);
+    if (onfinish) {
```
Comment (tchaikov) [inline src/mon/MonClient.cc]: note for posterity, we do have callers who don't pass a valid onfinish to `MonClient::start_mon_command()`, for instance, see https://github.com/ceph/ceph/blob/master/src/mgr/DaemonServer.cc#L2628

File: src/mon/MonClient.h [hunk 2]
```diff
@@ -501,6 +501,7 @@ class MonClient : public Dispatcher,
   struct MonCommand {
     std::string target_name;
     int target_rank;
+    int send_attempts = 0;
```
Comment (tchaikov) [inline src/mon/MonClient.h]: nit, use `unsigned`.

File: src/common/options.cc [hunk 3]
```diff
@@ -2180,6 +2180,10 @@ std::vector<Option> get_global_options() {
     .set_default(1000)
     .set_description(""),
 
+    Option("mon_client_directed_command_retry", Option::TYPE_INT, Option::DEV)
```
Comment (tchaikov) [inline src/common/options.cc]: s/DEV/LEVEL_DEV/

```
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/15.0.0-3218-g14457ae/rpm/el7/BUILD/ceph-15.0.0-3218-g14457ae/src/common/options.cc:2203:75: error: 'DEV' is not a member of 'Option'
     Option("mon_client_directed_command_retry", Option::TYPE_INT, Option::DEV)
                                                                           ^~~
```

see https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2019-07-27-0944/14457ae3fcefef873c45e1cf6871630e79d8c69d/notcmalloc/160520/

**Comment by gregsfortytwo** [comment]
I think I fixed all the typos and it looks good to me and apparently looked good to @tchaikov before, so let's try getting this through testing again.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-08-25_16:39:25-rados-wip-kefu-testing-2019-08-25-2138-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-08-27_00:38:42-rados-wip-kefu-testing-2019-08-26-1528-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-08-27_06:58:14-rados-wip-kefu-testing-2019-08-27-1029-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-08-27_06:45:12-rgw-wip-kefu-testing-2019-08-27-1029-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2019-08-27_08:50:36-rados-wip-kefu-testing-2019-08-27-1029-distro-basic-smithi/

**Pull request merged**